### PR TITLE
Candybar update

### PR DESCRIPTION
**Pull request by TerryMathews**
Splits Candybar into two subprojects to allow config.qmk.fm to show both possible layouts.

## Types of Changes

- [ ] Core
- [ ] Bugfix
- [ ] New feature
- [X] Enhancement/optimization
- [X] Keyboard (addition or update)
- [ ] Keymap/layout/userspace (addition or update)
- [ ] Documentation

## Issues Fixed or Closed by This PR

* Splits Candybar project into lefty and righty subprojects so that config.qmk.fm will show both layouts and still allow for Defaultish layouts.

## Checklist

- [X] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [X] I have updated the documentation accordingly.
- [X] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [ ] I have added tests to cover my changes.
- [X] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

### CODE REVIEW

**Comment by tzarc** [comment]
Is this just to support showing default layouts on configurator?
I'd be quite reluctant to duplicate entire keyboard definitions.

**Comment by TerryMathews** [comment]
That is correct, however there are actually two board versions so this is probably the most correct way to handle the situation.

**Comment by drashna** [comment]
> That is correct, however there are actually two board versions so this is probably the most correct way to handle the situation.

I would disagree with that.  It's identical boards, but two different entries in the repo, so it will work with the configurator.  That's ... a hack if I ever heard one. 

I do, however, understand the desire to have this working correctly. However, I feel that this this is the wrong way to go about this, and merging it sets a bad precedent.   

Ideally, this is something that should be fixed/improved on the configurator side, rather than here, as this basically artificially bloats the number of keyboards in the repo.  

Unfortunately, I don't have a good alternative to this, in the meanwhile.

**Comment by TerryMathews** [comment]
So where do we go from here? Fixing config to recognize default keymaps for every layout is well over my skill level.

I promise if config gains the ability to do that, that I will roll Candybar back to the way it exists today.

**Comment by drashna** [comment]
Terry,  I'm not exactly sure what is the best corse of action here.   Sory for not making it clear, byt my comment was more so it was "on the record", and as an explanation is why I I haven't put my approval on the PR.

From a code standpoint, it looks fine.  However, I think this is an issue that needs to be fixed on the configurator backen.  And not by you specifically.  

@yanfali @Erovia Thoughts?

**Comment by noroadsleft** [comment]
I actually have a stale branch of Configurator that allows setting a default keymap for each layout macro. @yanfali initially wasn't enthused about the implementation, and the fact of the matter is that I have a bit of a hard time maintaining all the keymaps that are already there (about 700 of them).

We do have an open issue of being able to make Configurator keymaps shareable via URL, in which case we could keep the system more or less as it is, and vendors could host their own Configurator keymap JSON files, and link them via the share URL.

https://github.com/qmk/qmk_configurator/issues/606

**Comment by TerryMathews** [comment]
My $0.02 is that the long-term solution for this is to re-write that part of configurator so that it properly greps the keymaps from the keymap folder from the QMK Master branch. Forget the defaultish solution, have it properly parse the actual keymap people are uploading.

Again, way over my skill level to implement. In the meantime, some solution would be appreciated even if it's a short-term fix. For some reason, it really bothers people to map their left-hand Candybar using the right-hand view even though it works perfectly.

**Comment by skullydazed** [comment]
> My $0.02 is that the long-term solution for this is to re-write that part of configurator so that it properly greps the keymaps from the keymap folder from the QMK Master branch. 

I actually just removed the keymap parsing from the API because it's so problematic and was not useful. Long term the idea is to move default keymaps to JSON and then configurator will be able to fetch them directly. I'll be providing a list of available JSON keymaps in the API to facilitate this. A major step in accomplishing that goal was recently merged and will be deployed shortly.

**Comment by TerryMathews** [comment]
So it seems the consensus is this PR won't be merged? How do I need to structure the code to meet whatever format you will allow to happen? One subproject, two layouts?

**Comment by noroadsleft** [comment]
This is probably something we need to solve on the QMK Configurator and API side of the equation. The QMK Firmware codebase is working as intended, so the problem isn't there – the limitation is Configurator being able to support only one keymap per keyboard revision at this time.

**Comment by TerryMathews** [comment]
Right, but this isn't that. These are separate PCBs, the matrix is different as is the keymap.

Is there no possibility of this getting merged until the future improvement to configurator is ready?

**Comment by noroadsleft** [comment]
Oh okay, I missed that there are actually two different PCBs, rather than one PCB that supports both layouts.

This is another HS60 situation (#7847) where a board that should have been submitted as more than one revision was not. Moving the build targets like this would potentially break keymap exports for anyone who uses Configurator, and may be a breaking change.

I'll review this later tonight. I need to see what the impact is.

**Comment by TerryMathews** [comment]
Yeah, there's two variants that physically move the 10key.

File: keyboards/candybar/righty/readme.md [hunk 1]
```diff
@@ -0,0 +1,18 @@
+The Key Company Candybar
+===
```
Comment (noroadsleft) [inline keyboards/candybar/righty/readme.md]:
```suggestion
# The Key Company Candybar
```

File: keyboards/candybar/righty/readme.md [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+
+![Candybar](https://cdn.shopify.com/s/files/1/1679/2319/articles/CandyBar_Promo_400x225_1000x.jpg?v=1538150501)
+
+
```
Comment (noroadsleft) [inline keyboards/candybar/righty/readme.md]:
```suggestion
```

File: keyboards/candybar/righty/readme.md [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+The Key Company Candybar is a staggered 40% board with a numpad utilizing the STM32F072 microcontroller.
+
+Keyboard Maintainer: [Terry Mathews](https://github.com/TerryMathews/)  
+Hardware Supported: TKC Candybar 
+Hardware Availability: Via GB
```
Comment (noroadsleft) [inline keyboards/candybar/righty/readme.md]:
```suggestion
* Keyboard Maintainer: [Terry Mathews](https://github.com/TerryMathews/)
* Hardware Supported: TKC Candybar
* Hardware Availability: [TheKey.Company](https://thekey.company/collections/candybar)
```

File: keyboards/candybar/righty/readme.md [hunk 4]
```diff
@@ -0,0 +1,18 @@
...
+![Candybar](https://cdn.shopify.com/s/files/1/1679/2319/articles/CandyBar_Promo_400x225_1000x.jpg?v=1538150501)
+
+
+The Key Company Candybar is a staggered 40% board with a numpad utilizing the STM32F072 microcontroller.
+
+Keyboard Maintainer: [Terry Mathews](https://github.com/TerryMathews/)  
+Hardware Supported: TKC Candybar 
+Hardware Availability: Via GB  
+
+
```
Comment (noroadsleft) [inline keyboards/candybar/righty/readme.md]:
```suggestion
```

File: keyboards/candybar/righty/righty.c [hunk 5]
```diff
@@ -0,0 +1,21 @@
...
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include "righty.h"
+
+void matrix_init_kb(void) {
+  matrix_init_user();
+}
```
Comment (noroadsleft) [inline keyboards/candybar/righty/righty.c]:
```suggestion
```

**Comment by TerryMathews** [comment]
Requested changes made including moving STM32 stubs into project base directory - which may have to be revisited if we ever run a future revision of Candybar on a different MCU but it's fine for today.

**Comment by tzarc** [comment]
I have plans to consolidate a large amount of the MCU stuff, removing them from keyboards so as to facilitate simpler ChibiOS upgrades in the future. With any luck that won't matter.

**Comment by TerryMathews** [comment]
@tzarc And I can say we're not doing anything special with our MCU code aside from making the soft reboot work correctly, which I understand is the same fix everyone has to use?

When the backend is ready, removing that code from this project will be easy. :)

**Comment by tzarc** [comment]
`__early_init` and `boardInit`? Yep, got #8330 in the pipeline for that, too.
https://github.com/qmk/qmk_firmware/blob/e3b67817779f3dda607834c881d703b26056700f/docs/platformdev_chibios_earlyinit.md

**Comment by tzarc** [comment]
Ideally, if #8330 gets merged you can do away with the board definitions in its entirety, and use the overridable hardware initialisation functions instead.

#8330 attempts to address the `__early_init` and `boardInit` problem
#8327 attempts to address the chibios conf files problem

**Comment by TerryMathews** [comment]
`void __early_init(void) {
  extern void enter_bootloader_mode_if_requested(void);
  enter_bootloader_mode_if_requested();
  stm32_gpio_init();
  stm32_clock_init();
}`

I believe this was the only change we made to the "base" file.

**Comment by tzarc** [comment]
There are mods in the `boardInit` which I'm guessing were duped from CannonKeys' F072 board defs.
Either way, should be covered :)

**Comment by zvecr** [comment]
As a FYI In this scenario, the boards folder is flat out not required since #8134. You can remove the files and `BOARD = ST_STM32F072B_DISCOVERY` from rules (I only kept it the same in that PR to minimise change and not have to tag every user for validation).

**Comment by tzarc** [comment]
Unless the current remap stuff is required... but considering the previous response I'd gather you're probably correct.

**Comment by TerryMathews** [comment]
Compiles correctly with Boards directory removed.

File: keyboards/candybar/righty/readme.md [hunk 6]
```diff
@@ -0,0 +1,15 @@
...
+* Keyboard Maintainer: [Terry Mathews](https://github.com/TerryMathews/)
+* Hardware Supported: TKC Candybar
+* Hardware Availability: [TheKey.Company](https://thekey.company/collections/candybar)
+
+Make example for this keyboard (after setting up your build environment):
+
+    make candybar/righty:default:dfu-util
+
+See [build environment setup](https://docs.qmk.fm/build_environment_setup.html) then the [make instructions](https://docs.qmk.fm/make_instructions.html) for more information.
```
Comment (noroadsleft) [inline keyboards/candybar/righty/readme.md]:
```suggestion
See the [build environment setup](https://docs.qmk.fm/#/getting_started_build_tools) and the [make instructions](https://docs.qmk.fm/#/getting_started_make_guide) for more information. Brand new to QMK? Start with our [Complete Newbs Guide](https://docs.qmk.fm/#/newbs).
```

**Comment by TerryMathews** [comment]
Done.

**Comment by zvecr** [comment]
Thanks!

**Pull request merged**